pina/geometry/difference_domain.py

```diff
@@ -1,28 +1,142 @@
+        """Check if a point is inside the difference domain.
+
+        :param point: Point to be checked.
+        :type point: torch.Tensor   
+        :param bool check_border: If True, the border is considered inside.
+        :return: True if the point is inside the difference domain, False otherwise.
+        :rtype: bool
+        """
+        flag = 0
+        for geometry in self.geometries:
+            if geometry.is_inside(point, check_border):
+                flag += 1
+        return flag == 1
+
+    def sample(self, n, mode='random', variables='all'):
+        """Sample routine.
 
-        import torch
-        return LabelTensor(torch.cat(samples), labels=['x', 'y'])
+        :param n: Number of points to sample in the shape.
+        :type n: int
+        :param mode: Mode for sampling, defaults to 'random'.
+            Available modes include: random sampling, 'random'.
+        :type mode: str, optional
+        :param variables: pinn variable to be sampled, defaults to 'all'.
+        :type variables: str or list[str], optional
+
+        :Example:
+            # Create two Cartesian domains
+            >>> cartesian1 = CartesianDomain({'x': [0, 2], 'y': [0, 2]})
+            >>> cartesian2 = CartesianDomain({'x': [1, 3], 'y': [1, 3]})
+
+            # Create a difference of the ellipsoid domains
+            >>> difference = Difference([cartesian1, cartesian2])
+
+            >>> difference.sample(n=1000)
+                LabelTensor([[1.6802, 0.7723],
+                    [1.8872, 0.9972],
+                    [0.3166, 0.1213],
+                    ...,
+                    [2.1999, 2.0767],
+                    [1.0655, 2.7711],
+                    [1.6751, 2.3695]])
+
+            >>> len(difference.sample(n=1000)
+                1000
+
+        """
+        sampled = []
+
+        # calculate the number of points to sample for each geometry and the remainder.
+        remainder = n % len(self.geometries)
+        num_points = n // len(self.geometries)
+
+        # sample the points
+        # NB. geometries as shuffled since if we sample
+        # multiple times just one point, we would end
+        # up sampling only from the first geometry.
+        iter_ = random.sample(self.geometries, len(self.geometries))
+        for i, geometry in enumerate(iter_):
+            sampled_points = []
+            # int(i < remainder) is one only if we have a remainder
+            # different than zero. Notice that len(geometries) is
+            # always smaller than remaider.
+            # makes sure point is uniquely inside 1 shape.
+            while len(sampled_points) < (num_points + int(i < remainder)):
+                sample = geometry.sample(1, 'random')
+                # if not self.is_inside(sample) --> will be the intersection
+                if self.is_inside(sample):
+                    sampled_points.append(sample)
+            sampled.extend(sampled_points)
```

Comment:
I believe `+=` is faster than `.extend()`. Let us check by doing a benchmark. You need to:

1. Sample [10^2, 10^3, 10^4, 10^5, 10^6] points from difference of two cartesians using `.extend()`
2. Do the same but use `+=` instead of `.extend()`
3. Plot times in the y-axis, and number of points in the `x.axis`. Use an appropriate scale of the axis (e.g. log-log, semi-log) for visualization and compare the two performances. 

Choose the one which takes less time and update the code. Also report the graph plot in the comments.

Reply:
![extend_vs_+=](https://github.com/mathLab/PINA/assets/54781152/cf9add2e-dd4d-432f-b077-ff303bf75888)

Reply:
Can you plot log-log scale?

Reply:
Do you still want this?

Reply:
![log_log](https://github.com/mathLab/PINA/assets/54781152/4b55c67f-5440-4ab5-9871-3184c9b2ee23)
Not sure that came out well. The numbers do show that `+=` is faster overall though.

Extend Data: `{'100': 0.08680009841918945, '1000': 0.5544390678405762, '10000': 5.0086829662323, '100000': 49.76351308822632, '1000000': 492.59749484062195}`

Plus Equals Data: `{'100': 0.047415971755981445, '1000': 0.4779529571533203, '10000': 4.886908292770386, '100000': 50.48829793930054, '1000000': 486.66690587997437}`